src/or/connection_edge.c

```diff
@@ -2767,6 +2767,15 @@ get_unique_stream_id_by_circ(origin_circuit_t *circ)
   for (tmpconn = circ->p_streams; tmpconn; tmpconn=tmpconn->next_stream)
     if (tmpconn->stream_id == test_stream_id)
       goto again;
+
+  if (circ->half_streams) {
+    SMARTLIST_FOREACH_BEGIN(circ->half_streams, half_edge_t *, half_conn) {
```

Comment:
Although I do wonder about the performance of this loop, every time we open a new stream.

Reply:
This loop will only run when we open a new stream on a circuit that had lots of half-closed streams. I think this is unlikely, and even worst case is only 65535 uint16_t compares, as per above. (Also we chose to use a O(N) search on the opened connections linked list for the same reasons).

Reply:
Ok, I decided to sort the list, so we can bsearch them on the common operations of cell arrival, and also guard against duplicate stream ids. I added tests for duplicates and list management correctness as well. I still think this is better than a hashtable, because for small list sizes, the memory allocation, hashing, and fragmentation overhead of a hashtable strikes me as not worth it.

Even on a full list of 65k half-closed connections (which should never happen), the slow operation (adding a new half-closed id, which triggers a sort), takes under 10ms on my 2015-era laptop. Since add only happens on EOF, even in this worst case, this will not directly affect user experience or page load.

7afb6f04e41236ebb6defb29fdd53cf9102bd925

Reply:
Ok sorry for the briandamage. I pushed a fixup to the original branch to use bsearch_idx at bcedf532945ba9d768dbdd6dfc1c9275ce829585

If you prefer, a commit of the two that is squashed is e60a239f61cca6efc8167b0ffa05c230af1e15a9

Reply:
Arg. I forgot to run the tests apparently. One more fixup: 60478b1c9627bf95fc2903c2e8caeda3159a584f 

The commit with all three of these fixups is 0b45fbda1fbbb1bf5785ba16c53375e59eb50a43